File: scripts/generate_content.py
"""
Affiliate content generator — produces Hugo-compatible Markdown via Groq API.
"""

import os
import json
from datetime import datetime
from pathlib import Path

from dotenv import load_dotenv
from groq import Groq
from slugify import slugify

load_dotenv()

client = Groq(api_key=os.environ["GROQ_API_KEY"])

SITE_DIR = Path(__file__).parent.parent / "site"
TEMPLATE_PATH = Path(__file__).parent.parent / "templates" / "post_template.md"

SYSTEM_PROMPT = """You are an expert affiliate marketing content writer targeting a global English-speaking audience.
Your goal: write SEO-optimized, conversion-focused product review articles.
Rules:
- Write in clear, natural American English
- Use H2 headings to structure the article
- Present a balanced view: honest pros AND cons
- Include a strong call-to-action (CTA)
- Maintain a trustworthy, helpful tone — never hype
- Aim for readers who are comparison-shopping and close to a buying decision
- Always respond with valid JSON only, no markdown code fences
"""


def generate_post(product_name: str, affiliate_link: str, category: str, extra_context: str = "") -> dict:
    """Generate a full blog post for the given product; returns structured JSON."""

    user_prompt = f"""
Product: {product_name}
Category: {category}
Affiliate Link: {affiliate_link}
{f'Additional context: {extra_context}' if extra_context else ''}

Return a JSON object with exactly these keys:
{{
  "title": "Article title (under 60 characters, compelling for search)",
  "meta_description": "SEO meta description (150-160 characters, include product name)",
  "rating": <integer 1-10>,
  "tags": ["tag1", "tag2", "tag3", "tag4"],
  "content": "Full article body in Markdown (minimum 600 words)"
}}

The content must include these sections in order:
- Introductory paragraph (hook the reader, state what this review covers)
- ## Key Features
- ## Pros and Cons
- ## Who Is It Best For?
- ## Price and Value
- ## Final Verdict
"""

    response = client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": user_prompt},
        ],
        temperature=0.7,
        response_format={"type": "json_object"},
    )

    data = json.loads(response.choices[0].message.content)
    data["product_name"] = product_name
    data["affiliate_link"] = affiliate_link
    data["categories"] = category
    # Consistent placeholder image keyed on product slug
    from slugify import slugify as _sl
    _slug = _sl(product_name)
    data["image"] = f"https://picsum.photos/seed/{_slug}/800/450"
    return data


def save_post(data: dict, content_type: str = "reviews") -> Path:
    """Save generated content as a Hugo-compatible Markdown file."""

    template = TEMPLATE_PATH.read_text(encoding="utf-8")

    date_str = datetime.now().strftime("%Y-%m-%dT%H:%M:%S+00:00")
    tags_str = ", ".join(f'"{t}"' for t in data.get("tags", []))

    post_content = template.format(
        title=data["title"],
        date=date_str,
        categories=f'"{data["categories"]}"',
        tags=tags_str,
        affiliate_link=data["affiliate_link"],
        product_name=data["product_name"],
        rating=data["rating"],
        image=data.get("image", ""),
        meta_description=data["meta_description"],
        content=data["content"],
    )

    filename = f"{datetime.now().strftime('%Y-%m-%d')}-{slugify(data['product_name'])}.md"
    output_dir = SITE_DIR / "content" / content_type
    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / filename

    output_path.write_text(post_content, encoding="utf-8")
    print(f"[OK] Saved: {output_path}")
    return output_path


def batch_generate(products: list[dict]) -> list[Path]:
    """Generate content for multiple products."""
    paths = []
    for i, product in enumerate(products, 1):
        print(f"\n[{i}/{len(products)}] Processing: {product['name']}")
        try:
            data = generate_post(
                product_name=product["name"],
                affiliate_link=product["link"],
                category=product.get("category", "general"),
                extra_context=product.get("context", ""),
            )
            path = save_post(data, content_type=product.get("type", "reviews"))
            paths.append(path)
        except Exception as e:
            print(f"[ERROR] {product['name']}: {e}")
    return paths


if __name__ == "__main__":
    sample_products = [
        {
            "name": "Xiaomi Redmi Note 13 Pro",
            "link": "https://amzn.to/example1",
            "category": "smartphones",
            "type": "reviews",
            "context": "Mid-range, AMOLED display, 200MP camera, 5000mAh battery, 67W fast charging",
        },
        {
            "name": "Logitech MX Master 3S",
            "link": "https://amzn.to/example2",
            "category": "computer-accessories",
            "type": "reviews",
            "context": "Silent clicks, ergonomic design, multi-device support, MagSpeed scroll wheel, 70-day battery",
        },
    ]

    saved = batch_generate(sample_products)
    print(f"\nDone — {len(saved)} articles generated.")

File: scripts/generate-content.ts
#!/usr/bin/env tsx
/**
 * AI Content Generation Script — DeskSetupLab
 * ================================================
 * Generates SEO-optimized product reviews and best-of pages
 * using OpenAI GPT-4o or Google Gemini.
 *
 * Usage:
 *   npx tsx scripts/generate-content.ts --type review --product flexispot-e7
 *   npx tsx scripts/generate-content.ts --type best --category standing-desks
 *   npx tsx scripts/generate-content.ts --type blog --topic "standing desk benefits"
 *
 * Env vars required:
 *   OPENAI_API_KEY or GEMINI_API_KEY
 */

import * as fs from "fs";
import * as path from "path";

// ── CLI Args ─────────────────────────────────────────────
const args = process.argv.slice(2);
const getArg = (name: string) => {
  const idx = args.indexOf(`--${name}`);
  return idx !== -1 ? args[idx + 1] : undefined;
};

const type = getArg("type") as "review" | "best" | "blog" | "comparison";
const topic = getArg("topic") ?? getArg("product") ?? getArg("category") ?? "";
const provider = (getArg("provider") ?? "openai") as "openai" | "gemini";
const dryRun = args.includes("--dry-run");

// ── AI Client Factory ─────────────────────────────────────
async function generateWithOpenAI(prompt: string): Promise<string> {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) throw new Error("OPENAI_API_KEY not set");

  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${apiKey}` },
    body: JSON.stringify({
      model: "gpt-4o",
      messages: [
        {
          role: "system",
          content:
            "You are an expert home office product reviewer. Write detailed, honest, SEO-optimized content in MDX format. Always include proper headings, real specs, pros/cons, and buying advice. Maintain journalistic integrity — never oversell.",
        },
        { role: "user", content: prompt },
      ],
      temperature: 0.7,
      max_tokens: 4000,
    }),
  });

  const data = await res.json();
  return data.choices?.[0]?.message?.content ?? "";
}

async function generateWithGemini(prompt: string): Promise<string> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) throw new Error("GEMINI_API_KEY not set");

  const res = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro:generateContent?key=${apiKey}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: [{ parts: [{ text: prompt }] }],
        generationConfig: { temperature: 0.7, maxOutputTokens: 4000 },
      }),
    }
  );

  const data = await res.json();
  return data.candidates?.[0]?.content?.parts?.[0]?.text ?? "";
}

async function generate(prompt: string): Promise<string> {
  return provider === "gemini" ? generateWithGemini(prompt) : generateWithOpenAI(prompt);
}

// ── Prompt Templates ──────────────────────────────────────
function reviewPrompt(productSlug: string): string {
  return `
Write a complete, hands-on product review for the home office product: "${productSlug}"

Format: MDX with frontmatter

Requirements:
- Title: "[Product Name] Review (${new Date().getFullYear()}) — Is It Worth It?"
- H2 sections: Overview, Key Features, Performance, Pros & Cons, Who It's For, Alternatives, Verdict
- Include real specs, dimensions, and pricing
- 1,500–2,000 words
- SEO-optimized: include keywords naturally
- Include Amazon affiliate disclosure note
- End with a clear buy/skip recommendation

Frontmatter fields:
- title, metaTitle, metaDescription, publishedAt, updatedAt, author, category, score (0-100), verdict
`;
}

function bestOfPrompt(category: string): string {
  return `
Write a "Best ${category} of ${new Date().getFullYear()}" article in MDX format.

Requirements:
- H1: "Best ${category} of ${new Date().getFullYear()} — Tested & Ranked"
- Intro paragraph (150 words): Why this category matters, how we tested
- List 5–7 products with: badge, name, price, pros, cons, who it's for, verdict
- H2 sections for each product: "1. [Product Name] — Best Overall", etc.
- Buying Guide section (300 words): 4 key things to look for
- FAQ section: 4 questions with detailed answers
- 2,500–3,000 words total
- Include affiliate disclosure
- Internal links to /review/[slug] pages

SEO requirements:
- Primary keyword: "best ${category.toLowerCase()}"
- Include LSI keywords naturally
`;
}

function blogPrompt(topic: string): string {
  return `
Write a blog post about: "${topic}" for a home office product review site.

Requirements:
- Informational/educational tone
- 1,000–1,500 words
- MDX format with frontmatter
- H2 and H3 structure
- Include actionable tips
- Naturally mention related products where relevant
- Include affiliate disclosure

Frontmatter: title, metaTitle, metaDescription, tags, category, publishedAt
`;
}

// ── Output Writer ─────────────────────────────────────────
function getOutputPath(contentType: string, slug: string): string {
  const dirMap: Record<string, string> = {
    review: "content/reviews",
    best: "content/best",
    blog: "content/blog",
    comparison: "content/comparisons",
  };
  const dir = path.join(process.cwd(), dirMap[contentType] ?? "content");
  fs.mkdirSync(dir, { recursive: true });
  return path.join(dir, `${slug}.mdx`);
}

// ── Main ──────────────────────────────────────────────────
async function main() {
  if (!type || !topic) {
    console.error("Usage: tsx generate-content.ts --type [review|best|blog] --product/category/topic [name]");
    process.exit(1);
  }

  console.log(`\n🤖 Generating ${type} content for "${topic}" using ${provider}...`);

  let prompt = "";
  switch (type) {
    case "review":
      prompt = reviewPrompt(topic);
      break;
    case "best":
      prompt = bestOfPrompt(topic);
      break;
    case "blog":
      prompt = blogPrompt(topic);
      break;
    default:
      console.error("Unknown type:", type);
      process.exit(1);
  }

  if (dryRun) {
    console.log("\n--- PROMPT PREVIEW ---");
    console.log(prompt);
    console.log("--- END PROMPT ---\n");
    console.log("Dry run complete. Remove --dry-run to generate content.");
    return;
  }

  try {
    const content = await generate(prompt);
    const slug = topic.toLowerCase().replace(/\s+/g, "-").replace(/[^\w-]/g, "");
    const outputPath = getOutputPath(type, slug);

    fs.writeFileSync(outputPath, content, "utf8");
    console.log(`\n✅ Content written to: ${outputPath}`);
    console.log(`📄 Word count: ~${content.split(/\s+/).length}`);
  } catch (err) {
    console.error("\n❌ Generation failed:", err);
    process.exit(1);
  }
}

main();
